### PR TITLE
WIP: auth: implement SSO using OIDC

### PR DESCRIPTION
**Pull request by mpldr**
This adds a basic implementation of SSO using OIDC.

It is missing:
- UI
- Documentation
- Feedback

Providing any of these would be much appreciated.

### CODE REVIEW

**Comment by mpldr** [comment]
Note: I do not feel comfortable doing the UI part, so I would request assistance on that part.

**Comment by m90** [comment]
Thanks for opening this PR. From reading the code alone I am a bit lost on what exactly this is trying to achieve. Could you add a few sentences about the "user story" this is trying to cover? This will also help us flesh out ideas around UI and docs and all of that.

**Comment by mpldr** [comment]
In order to enhance the user experience of offen when an OIDC provider
is available by implementing Single Sign-On (SSO) with OpenID Connect
(OIDC), this way users can seamlessly log in with their existing
credentials without the need for multiple logins. Especially in
commercial deployments, Domain Controllers or other forms of SSO like
Keycloak are rather common.

From a users perspective:
- visit offen.offen.dev/login/
- click on "Login with SSO"-Button
- they are redirected to the issuer (for example a Keycloak server)
- they log in with their existing credentials
- they are redirected back to offen
- their account is provisioned if it does not exist already
- they are logged in

From an operator's perspective:
- add required values (well-known-address, client id, client secret)
- offen now starts in "SSO"-mode, which entails the following:
	- no login with username and password
	- no password reset
	- if someone logs in with an email that does not have an
	  account, an account is created transparently

Excuse if this is a bit rough, my last proper user story was a few months ago.

-- 
Moritz Poldrack
https://moritz.sh

> This space intentionally left blank.

**Comment by m90** [comment]
> Excuse if this is a bit rough, my last proper user story was a few months ago.

Haha, no worries, this is perfect already and helps a lot. I'll have a look and think about it a bit. It's a welcome feature in any case.

File: server/go.mod [hunk 1]
```diff
@@ -46,6 +46,11 @@ require (
 	gorm.io/gorm v1.21.15
 )
 
+require (
+	golang.org/x/oauth2 v0.15.0 // indirect
+	mpldr.codes/oidc v0.0.0-20231223203712-a59dee5fc440
```
Comment (m90) [inline server/go.mod:51]: If I understand it correctly, this package is both authored and hosted by you. I don't mind this, but I'd be interested in learning more about:

- why you chose to write such a package when others exist
- what happens if your server goes away and this application still references the package using your domain. Does the Go package proxy kick in?

File: server/router/router.go [hunk 2]
```diff
@@ -233,16 +234,22 @@ func New(opts ...Config) http.Handler {
...
+
+			api.POST("/change-password", accountAuth, rt.postChangePassword)
+			api.POST("/change-email", accountAuth, rt.postChangeEmail)
+			api.POST("/forgot-password", rt.postForgotPassword)
+			api.POST("/reset-password", rt.postResetPassword)
+			api.POST("/share-account/:accountID", accountAuth, rt.postShareAccount)
+			api.POST("/share-account", accountAuth, rt.postShareAccount)
+			api.POST("/join", rt.postJoin)
+		} else {
+			api.POST("/login", rt.oauthLogin)
```
Comment (m90) [inline server/router/router.go:255]: Would there be any harm in always exposing these routes in addition to the existing ones and just 40x  out when someone calls them and OIDC is not configured? I would assume the URLs could be namespaced like `/oidc/login` or something.

File: server/persistence/login.go [hunk 3]
```diff
@@ -14,6 +15,22 @@ import (
 	"github.com/offen/offen/server/keys"
 )
 
+func (p *persistenceLayer) LoginSSO(email, salt string) (LoginResult, error) {
```
Comment (m90) [inline server/persistence/login.go:18]: The method receives a salt parameter, but it's not used from what I understand. Did you intend to use it when hashing the given email?

File: server/go.mod (hunk 1 above)
Comment (mpldr) [reply server/go.mod:51]: The package proxy kicks in. Worst case scenario is that there has to be a replace directive in the go.mod that points to the backing repo over at https://git.sr.ht/~mpldr/oidc

The main reason for writing it was to make OIDC stupid easy to use, since I've run into it a few times now. OIDC is basically oauth2, but the token contains extra details by default and the plan is to expose them in an easy to use manner.
Basically: you shouldn't have to understand how OIDC is implemented, what the tokens are signed with, and all that fun stuff.

File: server/persistence/login.go (hunk 3 above)
Comment (mpldr) [reply server/persistence/login.go:18]: That is indeed correct. I'll fix it in a bit :)

File: server/router/router.go (hunk 2 above)
Comment (mpldr) [reply server/router/router.go:255]: Sure, namespacing them would be easy. My main thought was: if the operator has an issuer set up, they likely won't use the internal login anyway, so there's no need for the form to be shown. One of my thoughts regarding the UI part was that it might be the easiest option to just use the "Login as operator" button as the "use SSO" button, which is why it's right now colliding.

Comment (m90) [reply server/router/router.go:255]: Hmm, then maybe let's defer the decision here until we know what the UI will look like. I'll talk about this with @hendr-ik beginning of next year and we'll feedback on how to proceed here.

**Comment by m90** [comment]
Currently, an `AccountUser` as stored in the database has several properties like an AdminLevel and AccountUserRelationships, tying the user to the accounts they are able to access.

How would this work when using OIDC? Would someone who logs in the first time still create a `AccountUser` record that can then be used to describe their access?

File: server/persistence/login.go [hunk 4]
```diff
@@ -14,6 +15,22 @@ import (
...
+	sha512Hash := sha512.New()
+	sha512Hash.Write([]byte(email))
+	sha512Hash.Write([]byte(salt))
+	dummyPassword := base64.URLEncoding.EncodeToString(sha512Hash.Sum(nil))
+	_, err := p.findAccountUser(email, false, false)
+	if err != nil {
```
Comment (m90) [inline server/persistence/login.go]: This should probably be more specific and only call `Join` when the error signals that no record was found. Joining should _not_ happen when for example network problem caused the query to fail.

File: server/persistence/login.go [hunk 5]
```diff
@@ -14,6 +15,22 @@ import (
 	"github.com/offen/offen/server/keys"
 )
 
+func (p *persistenceLayer) LoginSSO(email, salt string) (LoginResult, error) {
+	sha512Hash := sha512.New()
+	sha512Hash.Write([]byte(email))
+	sha512Hash.Write([]byte(salt))
```
Comment (m90) [inline server/persistence/login.go:21]: `Write` could potentially return an error, so this should be checked and handled as well.

Comment (mpldr) [reply server/persistence/login.go:21]: It can not. see `crypto/sha512/sha512.go:256`

File: server/persistence/login.go (hunk 4 above)
Comment (mpldr) [reply server/persistence/login.go]: Fixed in the latest version

**Comment by mpldr** [comment]
Thank you for the review. In OIDC something like this is returned:
```json
{
    "iss": "https://id.moritz.sh/application/o/codimd/",
    "sub": "1af2deecbcff74e06f140b10373d0907f3436f0c9451fe24df5c5e5963ebeb29",
    "aud": "b7bFpUvCVbUN00eNCFrNI6dyaYpCJcwuWT2gtCzm",
    "exp": 1704413169,
    "iat": 1704411369,
    "auth_time": 1704411369,
    "acr": "goauthentik.io/providers/oauth2/default",
    "amr": [
        "pwd",
        "mfa"
    ],
    "email": "moritz@poldrack.dev",
    "email_verified": true,
    "name": "Moritz Poldrack",
    "given_name": "Moritz Poldrack",
    "preferred_username": "moritz",
    "nickname": "moritz",
    "groups": [
        "admin",
        "users",
        "vpn",
    ]
}
```

Custom properties are of course possible, but the most reasonable approach to me seems to be using groups for this. If someone logs in for the first time, a "default" account is created for them. If, for example, a user was a member of the `admin` group, they should then get extended permissions. Regarding the `AccountUserRelationships`, this would be handled as it normally would, since OIDC creates a normal account. It's just that the possibilities for a login are limited.

File: server/persistence/login.go (hunk 5 above)
Comment (m90) [reply server/persistence/login.go:21]: Even if the current implementation does never assign a value to `err`, the signature still tells us it could do so at any time when someone changes the code.

I think it makes sense to treat the (stable) signature as the contract, not the (unstable) current implementation.

I do know this is a merely theoretical concern, but I think it makes sense to check errors if the signature tells us an error could be returned. Else, you'd need to start having intricate knowledge about all your dependencies and how they actually implement the things they are supposed to do.

**Comment by m90** [comment]
The interesting question probably is: should a deployment support users signing in with OIDC _and_ default user accounts, or should a deployment always use a single method. In case the latter is used, how does an operator configure to do what they want it to do?

I could see how the former is possible, but maybe it can get very confusing and is not a real world use case.

Another thing I was wondering: in case a deployment supports OIDC only, how is the first user account even created. Is it just the first person signing in using OIDC? Should there be a special workflow for doing this?

File: server/persistence/login.go (hunk 5 above)
Comment (mpldr) [reply server/persistence/login.go:21]: While I agree in principle, in this case the only reason it would fail was if there were issues in writing to memory. And if that happens, you have bigger issues. The error return is simply part of the `io.Writer` interface, and thus required to satisfy it. Hashing static data isn't something that can "fail".

**Comment by mpldr** [comment]
> should a deployment support users signing in with OIDC and default user accounts, or should a deployment always use a single method

I would argue that it's the latter. That's pretty much the entire point. This would also allow getting rid of all kinds of potential edge-cases in user interactions.

> in case a deployment supports OIDC only, how is the first user account even created. Is it just the first person signing in using OIDC?

Yes. Ideally, the first user shouldn't be anything special. Going by depolyment reality, it would likely be an admin anyway.

**Comment by m90** [comment]
> I would argue that it's the latter. That's pretty much the entire point. This would also allow getting rid of all kinds of potential edge-cases in user interactions.

I agree. In that case, however we might need to come up with a mechanism that "locks" a deployment to a single method of sign in and prevents it from being changed in the lifetime of such a deployment. I'll think about this for a bit, but if you have any ideas on the topic, feel free to voice them here.

**Comment by m90** [comment]
So today I met with @hendr-ik and we talked about how to get this feature shipped. The plan would be to complete the following steps:

## Server side functionality

This chunk of work would need to be finished first, ideally by you @mpldr 

### Delegate Login to OIDC provider

This is what you have already implemented here, so that's probably to be considered done.

### Devise mechanism that locks a deployment to  a certain method of login

As previously discussed in this PR, it would make sense to lock a deployment to a single method of managing users. Once a deployment has users that have been created using the default method, it should not be possible to change the login mechanism anymore and vice versa.

On a technical level, this could look like this: a `type` (or something similar) field is added to the `AccountUser` model. This holds either `default` or `oidc`. Each time a new account user is created, the deployment checks whether existing users are of the same type. In case they aren't, account creation is rejected. This means the first user created determines which method of login is used for a deployment. In addition to that it would be possible to check configuration on startup and try to match it with the users that exist in the database. In case there is a mismatch, the application would refuse to boot, which is probably helpful for users that inadvertently change their configuration to something unexpected.

### Add an endpoint to the server that lets the UI read "capabilities"

The Auditorium will now need to know which login screen display to users based on the server configuration. This could be done in two ways:

- when rendering the HTML for the Auditorium, the server adds a `window.capabilities` object containing information about login methods and other possibly other things that the SPA can then read and act upon
- a `/capabilities` endpoint is added to  the JSON api which is then called once by the SPA

Based upon this, the screens for `/login` and `/setup`  can then choose what to display.

### Unit tests

New code should be accompanied by unit tests as far as this is possible.

## Update Auditorium

Once this is done, we can work on the additions for the user interface. This will probably be worked on by @hendr-ik 

### Additional option for `/login` and `/setup`

When OIDC is used, render a single "Log in with OIDC" button instead of the login/setup form.

### Do not render "change password" and "change email" fields in Admin console

These options make no sense when using OIDC and should be replaced by a message about how this deployment's account users are managed using OIDC.

### Update translation strings

Once all of this is done, we'll need to reach out to our translators and have the new strings translated.

## Update docs

This can be done in parallel to working on the Auditorium by @m90 .

We want to avoid pushing new users towards decision paralysis when trying out Offen, so we thought it makes sense to be low key about the new option and default to the current mechanism in all tutorials. In addition to that, a detailed article explains how OIDC can be used, including all of the caveats that come with it.

----

Let me know if this makes sense for you @mpldr and if you need any further feedback from our end to wrap up the inital chunk of work on this topic.

**Comment by m90** [comment]
Closing this as it's stale. In case you want to revive it, please feel free to reopen.

**Comment by mpldr** [comment]
I will likely be able to get back to it in May :)

**Comment by mpldr** [comment]
@m90 May has turned into June, I can't seem to reopen it, though. Would you mind doing the honors?

**Comment by mpldr** [comment]
> Once a deployment has users that have been created using the default method, it should not be possible to change the login mechanism anymore and vice versa.

The approach I've chosen is adding the subject to the account information. For performance-reasons, since the setup only happens on startup, this would require a restart, though. This seems like an unnecessary requirement to me, so what I would suggest is using SSO as a sort of "extension":

- "Default" is still a normal account
- the account can be linked to an SSO-subject, but that disables "normal" login by overriding the password with a secure autogenerated one
- Password-changes are then disabled. Accounts are identified by a non-empty subject field.
- Normal Login can be disabled using an environment variable

Thoughts welcome.

**Comment by m90** [comment]
> so what I would suggest is using SSO as a sort of "extension"

I'm not entirely sure I understand yet. Does this mean you could have SSO user accounts as well as password user accounts in parallel as long as password login isn't disabled in the config?

**Comment by mpldr** [comment]
On Fri 21 Jun 2024 00:51:02, Frederik Ring wrote:
> I'm not entirely sure I understand yet. Does this mean you could have SSO user accounts as well as password user accounts in parallel as long as password login isn't disabled in the config?

The main idea is providing a "migration path" for users who wish to
introduce SSO into an existing system, but it would also reduce the
complexity of the implementation itself.

-- 
Moritz Poldrack
https://moritz.sh

> Serving suggestion.

**Comment by m90** [comment]
Two things I still don't fully understand assuming OIDC is an "extension":

- how does creation of the first account on setup work in this scenario?
- when OIDC is not configured, what happens to UI and such?

**Comment by mpldr** [comment]
On Sun 23 Jun 2024 04:44:23, Frederik Ring wrote:
> - how does creation of the first account on setup work in this scenario?

As now. It can then be linked to a SSO subject

> - when OIDC is not configured, what happens to UI and such?

The signin-with-SSO button isn't shown and no option to link the account
is provided. As for how to achieve that, I would leave the details on
how to do this to someone who knows what they're doing UI-wise and would
happily assist in getting the data to the frontend.

-- 
Moritz Poldrack
https://moritz.sh

> If not completely satisfied, return for full refund of purchase price.

**Comment by m90** [comment]
> As now. It can then be linked to a SSO subject

I start to understand, and it makes sense to me yes. Is there anything you need to implement it in that fashion still?

**Comment by mpldr** [comment]
> Is there anything you need to implement it in that fashion still?

Mostly information on what the frontend would need/want.

**Comment by m90** [comment]
I would think most of what's written in here https://github.com/offen/offen/pull/816#issuecomment-1889707113 still applies, apart from the things that say "either OIDC or password-based" are now "password-based and OIDC if configured".
